src/Orleans.Runtime/MembershipService/ClusterHealthMonitor.cs

```diff
@@ -214,7 +230,17 @@ async Task PingSilo(SiloHealthMonitor monitor, CancellationToken pingCancellatio
                     return;
                 }
 
-                this.log.LogWarning("Silo {Silo} failed {FailedProbes} probes and is suspected of being dead. Publishing a death vote.", monitor.SiloAddress, failedProbes);
+                // If the local node's health is degraded, we want to prevent it from evicting other nodes.
+                // Otherwise, a cluster with a high proportion of unhealthy nodes can evict the remaining healthy nodes.
+                if (this.isLocalHealthDegraded())
```

Comment:
I wonder if it's too simplistic for a silo to do nothing is it deems itself being in a degraded state. Wouldn't this potentially lead to an impasse if all or enough silos make that assessment and abstain from voting other silos dead?  Should we instead consider suspending votes from a silo in degraded state for a bounded period of time, hoping that it would either recover or itself get declared dead over that period of time? If neither of these things happened over that 'grace period', we would be back to today's behavior that at least doesn't leave the cluster stuck in an unhealthy state.

Reply:
> Should we instead consider suspending votes from a silo in degraded state for a bounded period of time, hoping that it would either recover or itself get declared dead over that period of time?

Yes, I think something like this is a good idea. I was considering alternatively using the score to increase ping cancellation times instead, or increasing the number of ping failures required for a vote, which would have a similar effect.